File: Searching/Sahil/C++/binary_search.cpp
/*
Works only with sorted arrays
*/

#include<iostream>
#include<array>
using namespace std;

int binary_search(int arr[], int size, int number_to_search) {

    int min = 0 ;
    int max = size-1 ;
    int index = -1 ;

    while( min <= max ) {

        int mid = (max + min) / 2 ;
            
        if(arr[mid] < number_to_search) {
            min = mid + 1 ;
        } else if(arr[mid] > number_to_search) {
            max = mid - 1 ;
        } else {
            index = mid ;
            break;
        }
    }
    return index ;
}

int main() {

    array<int, 5> array = { 2, 3, 4, 10, 40 } ;

    int number_to_search = 89 ;

    int search_result = binary_search(array.data(), array.size(), number_to_search) ;

    if (search_result >= 0) {

        cout << "got it at: " << search_result + 1 << endl ;
    } else {

        cout << "No such value present" << endl ;
    }
    return 0 ;
}

File: Searching/Sahil/JAVA/BinarySearch.java
package searching;

public class BinarySearch {

    public static int binarySearch(int[] arr, int numberToSearch) {

        int position = -1 ;
        int low = 0, high = arr.length - 1 ;
        while (low <= high) {

            int mid = (high + low) / 2 ;

            if(arr[mid] < numberToSearch) {
                low = mid + 1 ;
            } else if(arr[mid] > numberToSearch) {
                high = mid - 1 ;
            } else {
                position = mid ;
                break ;
            }
        }
        return position ;
    }

    public static void main(String[] args) {

        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200 } ;
        int numberToSearch = 100 ;
        int searchResult = binarySearch(arr, numberToSearch) ;

        if(searchResult >= 0) {
            System.out.println("Number found at: " + (searchResult + 1));
        } else {
            System.out.println("Number not found");
        }
    }
}
